dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -754,58 +693,76 @@ void DMFileReader::addSkippedRows(UInt64 rows)
+                read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, last_pack_res, read_rows);
+            start_pack_id = pack_id;
+            read_rows = pack_stats[pack_id].rows;
+            last_pack_res = pack_res[pack_id];
+        }
+        else
+        {
+            last_pack_res = last_pack_res && pack_res[pack_id];
+            read_rows += pack_stats[pack_id].rows;
         }
-        auto [pack_count, rows] = get_all_match_block(i);
-        // Do not read block too small, it may hurts performance
-        if (rows >= rows_threshold_per_read / 2)
-            all_match_block_infos.emplace(i, pack_count);
-        i += pack_count;
     }
+    if (read_rows > 0)
+        read_block_infos.emplace_back(start_pack_id, pack_res.size() - start_pack_id, last_pack_res, read_rows);
 }
 
-size_t DMFileReader::getReadPackLimit(size_t start_pack_id)
+size_t DMFileReader::updateReadBlockInfos(const IColumn::Filter & filter, size_t pack_start, size_t pack_end)
 {
-    if (all_match_block_infos.empty() || read_one_pack_every_time)
-        return read_one_pack_every_time ? 1 : std::numeric_limits<size_t>::max();
-
-    const auto [next_all_match_block_start_pack_id, pack_count] = all_match_block_infos.front();
-    // Read packs with RSResult::All
-    if (next_all_match_block_start_pack_id == start_pack_id)
+    const auto & pack_stats = dmfile->getPackStats();
+    std::vector<std::tuple<size_t, size_t, RSResult, size_t>> new_read_block_infos;
+    new_read_block_infos.reserve(pack_end - pack_start);
+    size_t start_pack_id = pack_start;
+    const size_t start_row_offset = pack_id_to_offset[pack_start];
+    size_t read_rows = 0;
+    for (size_t pack_id = pack_start; pack_id < pack_end; ++pack_id)
     {
-        all_match_block_infos.pop();
-        return pack_count;
+        if (countBytesInFilter(filter, pack_id_to_offset[pack_id] - start_row_offset, pack_stats[pack_id].rows) == 0)
+        {
+            if (read_rows > 0)
+                // rs_result is not important here, we can use RSResult::All
+                new_read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, RSResult::All, read_rows);
+            start_pack_id = pack_id + 1;
+            read_rows = 0;
+        }
+        else
+        {
+            read_rows += pack_stats[pack_id].rows;
+        }
     }
-    // Read packs until next_all_match_block_start_pack_id
-    RUNTIME_CHECK(
-        next_all_match_block_start_pack_id > start_pack_id,
-        next_all_match_block_start_pack_id,
-        start_pack_id);
-    return next_all_match_block_start_pack_id - start_pack_id;
+    if (read_rows > 0)
+        new_read_block_infos.emplace_back(start_pack_id, pack_end - start_pack_id, RSResult::All, read_rows);
+    size_t size = new_read_block_infos.size();
+    for (auto & new_read_block_info : std::ranges::reverse_view(new_read_block_infos))
+        read_block_infos.emplace_front(std::move(new_read_block_info));
+    return size;
```

Comment:
Why emplace the `new_read_block_infos` into the front of `read_block_infos`? Can we just replace the `block_read_infos` by `new_read_block_infos`?

Reply:
read_block_infos: [0, 6), [6, 10), [10, 15)
filter out packs 2, 4
new_read_block_infos: [0, 1), [3, 4), [5, 6)
read_block_infos: [0, 1), [3, 4), [5, 6), [6, 10), [10, 15)

Reply:
So in `readWithFilter`, it break the first "read_block_infos" [0, 6) into smaller parts by the filter. What about moving `const auto [start_pack_id, pack_count, rs_result, read_rows] = read_block_infos.front(); read_block_infos.pop_front();` into `updateReadBlockInfos`?

Reply:
done